### PR TITLE
[CUDA] Fix AvgPool2d backward handling of zero-padding 

### LINKED ISSUE #188344: avg_pool2d backward returns wrong input gradients on CUDA for channels_last inputs when padding > 0
### 🐛 Describe the bug

`avg_pool2d` backward on CUDA produces incorrect input gradients when the input is in `channels_last` memory format and `padding > 0`. The forward output is correct, the `channels_last` result with `padding = 0` is correct, and the contiguous (NCHW) result is correct for any padding. Only the `channels_last` + `padding > 0` backward is wrong, and it fails silently (no error, just wrong numbers), so it can corrupt training gradients for any model that pools `channels_last` activations with padding.

Minimal repro (compares the eager `channels_last` gradient against the CPU result, which is the reference):

```python
import torch
import torch.nn.functional as F

def grad(device, channels_last):
    x = torch.arange(1 * 2 * 4 * 4, device=device, dtype=torch.float64).reshape(1, 2, 4, 4)
    if channels_last:
        x = x.contiguous(memory_format=torch.channels_last)
    x = x.detach().requires_grad_()
    out = F.avg_pool2d(x, kernel_size=2, stride=1, padding=1)
    out.backward(torch.ones_like(out))
    return x.grad.detach().cpu()

ref = grad("cpu", channels_last=False)
print("cuda NCHW          max|err|:", (grad("cuda", False) - ref).abs().max().item())
print("cuda channels_last max|err|:", (grad("cuda", True)  - ref).abs().max().item())
```

Output (RTX 4090, `2.14.0.dev20260622+cu132`):

```
cuda NCHW          max|err|: 0.0
cuda channels_last max|err|: 0.75
```

Expectation: both the NCHW and the `channels_last` CUDA backward should match the CPU reference. Actual: the `channels_last` backward is wrong whenever `padding > 0`.

The deviation is not numerical noise. It is a systematic spatial shift of the gradient, and it is independent of kernel size, stride, and `count_include_pad`. A scan over `kernel in {(2,2),(2,3),(3,3)}`, `stride in {1,2}`, `padding in {0,1}`, `count_include_pad in {True, False}` shows the `channels_last` backward is exact for `padding = 0` and wrong for `padding = 1` in every case, while the forward and the NCHW backward are always exact. `torch.compile(backend="inductor")` also produces the correct (NCHW/CPU-matching) gradient, so only the eager CUDA `channels_last` backward is affected.

I have a fix and a regression test ready and can send a PR.

### Versions

```
PyTorch version: 2.14.0.dev20260622+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Python version: 3.12.3 (64-bit runtime)
Is CUDA available: True
GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 580.105.08
cuDNN version: 9.8.0
```

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @ptrblck @eqy @tinglvv @nWEIdia @jamesr66a @mikaylagawarecki

### PR DESCRIPTION
**Pull request by eqy**
To fix #188344

Original test from https://github.com/pytorch/pytorch/pull/35855 didn't actually pass padding which is why we didn't see it

authored with codex

cc @ptrblck @msaroufim @tinglvv @nWEIdia

### CODE REVIEW

**Comment by drisspg** [comment]
@claude this fix looks correct and makes sense to me, my curious question is why wasnt this noticed for so long / curious if you have any insights into this specific setup as well im curious if you can do a wider code base search and see if there are any other problems of a similar form that we have missed

**Review by drisspg — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]

**Review by ngimel — APPROVED** [review]

**Comment by eqy** [comment]
https://github.com/pytorch/pytorch/pull/188345 was merged
